core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -272,15 +364,270 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+   * @return the truncated leader epoch checkpoint
+   */
+  private[remote] def getLeaderEpochCheckpoint(log: UnifiedLog, startOffset: Long, endOffset: Long): InMemoryLeaderEpochCheckpoint = {
+    val checkpoint = new InMemoryLeaderEpochCheckpoint()
+    log.leaderEpochCache
+      .map(cache => cache.writeTo(checkpoint))
+      .foreach { x =>
+        if (startOffset >= 0) {
+          x.truncateFromStart(startOffset)
+        }
+        x.truncateFromEnd(endOffset)
+      }
+    checkpoint
+  }
+
+  private[remote] class RLMTask(tpId: TopicIdPartition) extends CancellableRunnable with Logging {
+    this.logIdent = s"[RemoteLogManager=$brokerId partition=$tpId] "
+    @volatile private var leaderEpoch: Int = -1
+
+    private def isLeader(): Boolean = leaderEpoch >= 0
+
+    // The readOffset is None initially for a new leader RLMTask, and needs to be fetched inside the task's run() method.
+    @volatile private var copiedOffsetOption: Option[Long] = None
+
+    def convertToLeader(leaderEpochVal: Int): Unit = {
+      if (leaderEpochVal < 0) {
+        throw new KafkaException(s"leaderEpoch value for topic partition $tpId can not be negative")
+      }
+      if (this.leaderEpoch != leaderEpochVal) {
+        leaderEpoch = leaderEpochVal
+      }
+      // Reset readOffset, so that it is set in next run of RLMTask
+      copiedOffsetOption = None
+    }
+
+    def convertToFollower(): Unit = {
+      leaderEpoch = -1
+    }
+
+    def handleCopyLogSegmentsToRemote(): Unit = {
+      if (isCancelled())
+        return
+
+      def maybeUpdateReadOffset(): Unit = {
+        if (copiedOffsetOption.isEmpty) {
+          info(s"Find the highest remote offset for partition: $tpId after becoming leader, leaderEpoch: $leaderEpoch")
+
+          // This is found by traversing from the latest leader epoch from leader epoch history and find the highest offset
+          // of a segment with that epoch copied into remote storage. If it can not find an entry then it checks for the
+          // previous leader epoch till it finds an entry, If there are no entries till the earliest leader epoch in leader
+          // epoch cache then it starts copying the segments from the earliest epoch entry’s offset.
+          copiedOffsetOption = Some(findHighestRemoteOffset(tpId))
+        }
+      }
+
+      try {
+        maybeUpdateReadOffset()
+        val copiedOffset = copiedOffsetOption.get
+        fetchLog(tpId.topicPartition()).foreach { log =>
+          // LSO indicates the offset below are ready to be consumed(high-watermark or committed)
+          val lso = log.lastStableOffset
+          if (lso < 0) {
+            warn(s"lastStableOffset for partition $tpId is $lso, which should not be negative.")
+          } else if (lso > 0 && copiedOffset < lso) {
+            // Copy segments only till the min of high-watermark or stable-offset as remote storage should contain
+            // only committed/acked messages
+            val toOffset = lso
```

Comment:
Where do we set the offset with `min of high-watermark or stable-offset`?

Reply:
LastStableOffset already represents that, will update the javadoc comment.